src/cascade/input_data/emr.py

```diff
@@ -0,0 +1,219 @@
+    csmr = csmr.merge(mean_years, on=["age_group_id", "node_id", "sex_id", "time_lower"])
+    csmr["age"] += csmr["age_lower"]
+    return csmr
+
+
+def _prepare_csmr(execution_context, csmr):
+    MATHLOG.debug("Preparing CSMR data from GBD")
+    csmr = csmr.rename(columns={"location_id": "node_id"})
+
+    MATHLOG.debug("Assigning standard error from measured upper and lower.")
+    csmr = meas_bounds_to_stdev(csmr)
+
+    null_means = csmr["mean"].isnull()
+    if np.any(null_means):
+        MATHLOG.warn(
+            f"{np.sum(null_means)} CSMR rows with NaN means in EMR calculation. "
+            "Removing them but there may be something wrong"
+        )
+        csmr = csmr.loc[~null_means]
+
+    csmr = age_groups_to_ranges(execution_context, csmr, keep_age_group_id=True)
+    if execution_context.parameters.use_weighted_age_group_midpoints:
+        MATHLOG.debug("Treating CSMR measurements over age ranges as measurements at the mortality weighted midpoint")
+        csmr = _collapse_ages_weighted(execution_context, csmr)
+    else:
+        MATHLOG.debug("Treating CSMR measurements over age ranges as measurements at the unweighted midpoint")
+        csmr = _collapse_ages_unweighted(csmr)
+    csmr = csmr.drop(columns=["age_lower", "age_upper"])
+
+    csmr = _collapse_times(csmr)
+
+    return csmr
+
+
+def _make_interpolators(csmr):
+    """ Calculate interpolations over CSMR so we can integrate the CSMR value
+    over the age/time range of each prevalence measurement.
+    """
+    mean = {}
+    stderr = {}
+    mean["both"] = LSQBivariateSpline(
+        csmr.age, csmr.time, csmr["mean"], sorted(csmr.age.unique()), sorted(csmr.time.unique()), kx=1, ky=1
+    )
+    stderr["both"] = LSQBivariateSpline(
+        csmr.age, csmr.time, csmr.standard_error, sorted(csmr.age.unique()), sorted(csmr.time.unique()), kx=1, ky=1
+    )
+
+    # csmr_by_age = csmr.sort_values("age").groupby("age").mean().reset_index()
+    # mean["age"] = InterpolatedUnivariateSpline(csmr_by_age.age, csmr_by_age["mean"], k=1)
+    # stderr["age"] = InterpolatedUnivariateSpline(csmr_by_age.age, csmr_by_age.standard_error, k=1)
+
+    csmr_by_time = csmr.sort_values("time").groupby("time").mean().reset_index()
+    mean["time"] = InterpolatedUnivariateSpline(csmr_by_time.time, csmr_by_time["mean"], k=1)
+    stderr["time"] = InterpolatedUnivariateSpline(csmr_by_time.time, csmr_by_time.standard_error, k=1)
+
+    return mean, stderr
+
+
+def _emr_from_sex_and_node_specific_csmr_and_prevalence(csmr, prevalence):
+    mean_interp, stderr_interp = _make_interpolators(csmr)
+
+    emr = prevalence[["age_lower", "age_upper", "time_lower", "time_upper", "sex_id", "node_id", "density", "weight"]]
+    emr["measure"] = "mtexcess"
+
+    def emr_mean(prevalence_measurement):
+        if (
+            prevalence_measurement.time_lower == prevalence_measurement.time_upper
```

Comment:
Doing the equivalence test multiple times might be considered premature pessimization. What would you think of assigning a variable that says time_is_point_estimate or age_is_point_estimate? That way we could reuse those in the elif below.

Reply:
Sure